### PR TITLE
Fix for error protecting dashboards (Connect #2024)

### PR DESCRIPTION
**Pull request by kardan**
There are some outstanding issues. For example, if a password is set. And then the user tries to update the password to a too short one. He/she will get an error dialog at the bottom of the screen (the backend responded with error) but the alert in the modal will be all happy with a green succeded. But I think that is another issue than this initial one. I tried to fix that one too, but backed out of it since it looked to me like a little bit of rearchitecting is needed. @finnfiddle if you think different ping me.

### CODE REVIEW

File: client/src/containers/Dashboard.jsx [hunk 1]
```diff
@@ -432,8 +433,7 @@ class Dashboard extends Component {
 
   handleToggleShareProtected(isProtected) {
     this.setState({ passwordAlert: null });
-    const dashboard = getDashboardFromState(this.state.dashboard, true);
-    this.props.dispatch(actions.setShareProtection(dashboard.shareId, { protected: isProtected }));
+    set(this.state, 'dashboard.protected', isProtected);
```
Comment (finnfiddle) [inline client/src/containers/Dashboard.jsx]: this doesnt look right. Maybe:
```
this.setState({ 'dashboard.protected': isProtected });
```

Comment (finnfiddle) [reply client/src/containers/Dashboard.jsx]: or actually:
```
this.setState({ 
  dashboard: {
    ...this.state.dashboard,
    protected: isProtected,
  },
});

Comment (finnfiddle) [reply client/src/containers/Dashboard.jsx]: if you just set it manually (ie not using the `setState` method then the observers wont get triggered

File: client/src/containers/Dashboard.jsx [hunk 2]
```diff
@@ -432,8 +433,7 @@ class Dashboard extends Component {
 
   handleToggleShareProtected(isProtected) {
     this.setState({ passwordAlert: null });
-    const dashboard = getDashboardFromState(this.state.dashboard, true);
-    this.props.dispatch(actions.setShareProtection(dashboard.shareId, { protected: isProtected }));
```
Comment (finnfiddle) [inline client/src/containers/Dashboard.jsx]: also why are we removing this API request? Isnt this how we save the config?

Comment (kardan) [reply client/src/containers/Dashboard.jsx]: This get's called when we enable the radio button & yet have to type a password. The backend will respond with an error message (no password) and the client will see the error. This removes the initial request. And the user has to hit "save" once he/she entered a valid password. At that point a request with all config necessary to enable protection is available.

**Pull request merged**